yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +102,180 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     if (credentials != null) credentials.getSecretKey(new Text(key)) else null
   }
 
+  private[spark] override def scheduleLoginFromKeytab(): Unit = {
+    val principal = System.getenv("SPARK_PRINCIPAL")
+    val keytab = System.getenv("SPARK_KEYTAB")
+    if (principal != null) {
+      val delegationTokenRenewerThread =
+        new Runnable {
+          override def run(): Unit = {
+            if (!loggedInViaKeytab) {
+              // Keytab is copied by YARN to the working directory of the AM, so full path is
+              // not needed.
+              loggedInUGI = UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+                principal, keytab)
+              loggedInViaKeytab = true
+            }
+            val nns = getNameNodesToAccess(sparkConf)
+            val newCredentials = loggedInUGI.getCredentials
+            obtainTokensForNamenodes(nns, conf, newCredentials)
+            val remoteFs = FileSystem.get(conf)
+            val stagingDirPath =
+              new Path(remoteFs.getHomeDirectory, System.getenv("SPARK_YARN_STAGING_DIR"))
+            val tokenPathStr = sparkConf.get("spark.yarn.credentials.file")
+            val tokenPath = new Path(stagingDirPath.toString, tokenPathStr)
+            val tempTokenPath = new Path(stagingDirPath.toString, tokenPathStr + ".tmp")
+            val stream = remoteFs.create(tempTokenPath, true)
+            // Now write this out to HDFS
+            newCredentials.writeTokenStorageToStream(stream)
+            stream.hflush()
+            stream.close()
+            // HDFS does reads by inodes now, so just doing a rename should be fine. But I could
+            // not find a clear explanation of when the blocks on HDFS are deleted. Ideally, we
+            // would not need this, but just be defensive to ensure we don't mess up the
+            // credentials. So create a file to show that we are currently updating - if the
```

Comment:
That was what I planned to do, but I was not entirely sure about the file listing aspect of things - whether it would affect perf. But you are right - there are too many race conditions right now. We should avoid it.